src/index.js

```diff
@@ -432,6 +432,36 @@ class MiniCssExtractPlugin {
           return source;
         }
       );
+
+      compilation.hooks.optimizeChunks.tap(pluginName, (chunks) => {
+        for (const chunk of chunks) {
+          chunk.modulesIterable.forEach((module) => {
+            // eslint-disable-next-line no-param-reassign
+            module.dependencies = module.dependencies.filter((dep) => {
+              const hasCssDep =
+                dep.module &&
+                dep.module.dependencies.find(
```

Comment:
The whole loop is very expensive. `chunks * modules * dependencies * dependencies`. I don't think this is acceptable for larger builds.

Reply:
Definitely not performant. mini-css already does get expensive on large sites with many css files. It would be quite noticeable. 

What if we used something other than pure JS which has pretty bad performance on loops compared to lodash or transducers.js. 

https://jlongster.com/Transducers.js-Round-2-with-Benchmarks

@sokra I have been considering trying to improve loop performance. Any thoughts on an implementation like one mentioned in the article? Benchmarks are impressive with substantial performance over native js functions

Reply:
Made a stress test for this case. Ran three times with these settings: size 100, depth 80. The results are quite interesting:

`npm run test:stress`

```
Mode: development
Modules: 24002
Build time (without loop):  94944ms, 99758ms, 97380ms
Build time (with loop):     92237ms, 98391ms, 94587ms
Assets (without loop):      development.js 15.3 MiB
Assets (with loop):         development.js 6.47 MiB

Mode: production
Modules: 16001
Build time (without loop):  96744ms, 102081ms, 99210ms
Build time (with loop):     92744ms, 98898ms, 94731ms
Assets (without loop):      production.js 203 KiB
Assets (with loop):         production.js 951 bytes (hi!)
```

Reply:
@sokra @ScriptedAlchemy @evilebottnawi please take a look

Reply:
What about internally using some `Set` or `Map` to memoize previous results and eliminate duplicate evaluation. Would that not help with modules duplicated over multiple chunks and in particular with reducing the tracing for css dependencies?

Somewhat like e.g.

```js
compilation.hooks.optimizeChunks.tap(pluginName, (chunks) => {

  const modules = new Set();
  
  for (const chunk of chunks) {
    for (const module of chunk.modulesIterable) {
      modules.add(module);
      
      if (module.type !== MODULE_TYPE || !module.reasons) break;
      
      for (const reason of module.reasons) {
        if (!reason.module || !reason.module.buildMeta.extracted) break;
        chunk.removeModule(reason.module);
      }
    }
  }
  
  const cssDeps = new Map();
  
  modules.forEach((module) => {
    module.dependencies = module.dependencies.filter((dep) => {
      let hadDep = cssDeps.get(dep);
      if (hadDep == null) {
        cssDeps.set(dep, hadDep = checkForCssDep(dep));
        if (hadDep) { dep.disconnect(); }
      }
          
      return !hadDep;
    });
  });
  
  function checkForCssDep( dep ) {
    return dep.module && dep.module.dependencies.some(
      (d) => d.module && d.module.type === MODULE_TYPE
    );
  }
});
```

Reply:
@sokra this implementation is much more complicated than the original. But performance seems like the same.

The tricky part is that we need to disconnect HarmonyImportSideEffectDependency, but we have only CssDependency.

Should we use this or previous version?

```javascript
class SetMap extends Map {
  set(key, value) {
    return super.set(key, this.has(key) ? this.get(key).add(value) : new Set([value]));
  }
}

// ...

const toRemoveMap = new SetMap();

for (const chunk of chunks) {
  for (const module of chunk.modulesIterable) {
    if (module.type === MODULE_TYPE) {
      for (const cssModuleReason of module.reasons) {
        for (const reason of cssModuleReason.module.reasons) { // 💩
          toRemoveMap.set(reason.module || cssModuleReason.module, reason.dependency);
        }
        if (cssModuleReason.module && cssModuleReason.module.buildMeta.extracted) {
          chunk.removeModule(cssModuleReason.module);
        }
      }
    }
  }
}

for (const [module, set] of toRemoveMap) {
  module.dependencies = module.dependencies.filter(d => {
    if (set.has(d)) {
      d.disconnect();
      return false;
    }
    return true;
  });
}
```